[release/2.1] Update branding to 2.1.28

dougbu opened the pull request:
- update branding to 2.1.28
- bump `$(MicrosoftNETCoreAppPackageVersion)`
    - version flows into `$(MicrosoftNETCoreApp21PackageVersion)` automatically
    - cherry-picked from #30729 internal branch
- get SharedFx.UnitTests working after rebranding
    - skip `SharedFxTests.BaselineTest(...)` between rebranding and baseline updates
        - previous runtime is not available in this window
    - reduce retries in `RetryHelper`
        - download will usually fail repeatedly if it fails even once
    - previous maximum duration (105 min.) was greater than the job timeout
- avoid NETSDK1023 warnings
    - eng/targets/CSharp.Common.props adds packages removed from project
    - cherry-picked from #30729 internal branch
  nit: remove unused `GetTotalRetriesUsed()` method
- use Ubuntu 18.04 build agents

dougbu committed:
Get tests and archives working after rebranding
- skip `SharedFxTests.BaselineTest(...)` between rebranding and baseline updates
    - previous runtime is not available in this window
- reduce retries in `RetryHelper`
    - download will usually fail repeatedly if it fails even once
    - previous maximum duration (105 min.) was greater than the job timeout
- avoid NETSDK1023 warnings
    - eng/targets/CSharp.Common.props adds packages removed from project
- create archives based on one that actually exists between rebranding and baseline updates
- mostly cherry-picked from #30729 internal branch

nit: remove unused `GetTotalRetriesUsed()` method

dougbu committed:
Bump `$(MicrosoftNETCoreAppPackageVersion)`
- version flows into `$(MicrosoftNETCoreApp21PackageVersion)` automatically
- cherry-picked from #30729 internal branch

dougbu committed:
Stop using Ubuntu 14.04
- run `locale-gen` to avoid "setlocale: LC_ALL: cannot change locale" and similar
    - needs `locales` package
- move to still-supported Ubuntu 18.04 image
    - image needs slightly-newer `libicu60`, `clang` and `lldb` packages
    - rename Dockerfile to match
- cherry-picked from #30729 internal branch for Ubuntu image update

nit: set locale-related environment variables to match Python default

dougbu committed:
Avoid KoreBuild issues when using dockerbuild.sh
- avoid MSB4011 warnings (about repeated Directory.Build.targets imports)
    - place `$DOTNET_HOME` beside repo root, not under it
- cherry-picked from #30729 internal branch for image update

nits:
- use current Docker image for 2.1 alpine
- set `$LANGUAGE` with other locale-related environment variables

dougbu committed:
Add debug output to CDN tests

dougbu committed:
Improve the other CDN tests
- apply efe9b95b1444 to similar template test

dougbu committed:
Update branding to 2.1.28

dougbu committed:
Use Ubuntu 18.04 build agents
- set locale consistently on all platforms
    - default locale on newer agents is unloved `C.UTF-8`

TanayParikh commented on eng/PatchConfig.props:
```diff
@@ -100,4 +100,8 @@ Later on, this will be checked using this condition:
       @aspnet/signalr-protocol-msgpack;
     </PackagesInPatch>
   </PropertyGroup>
+  <PropertyGroup Condition=" '$(VersionPrefix)' == '2.1.28' ">
```
Above changes are for `2.1.26` and this is `.28` is that intentional?

TanayParikh replied:
Ah so this is the `AspNetCore` package version https://github.com/dotnet/aspnetcore/pull/31547/files#diff-e4bed5b736f205989dd4fdb6d78acfe9126577983d325d378ce91794d74e63c8R5

dougbu replied:
Yes, my changes in build/dependencies.props are just to use the packages that are already public in rolling and PR builds. We'll move to 2.1.27 when updating the baselines on Patch Tuesday.

TanayParikh reviewed (approved):
Overall LGTM, not sure if other changes are required to make this official.

dougbu commented:
Thanks @TanayParikh I'll get this in tomorrow, when the branches are slated to open.

dougbu commented:
@wtgodbe @JunTaoLuo not sure if the commits I just added are needed for 3.1 or 5.0. Decided they made sense here because the Docker images we were using were very old.

But, this bulks up the PR. Let me know if you'd prefer I did some parts in separate PRs.

dougbu commented:
Holding for a actions/virtual-environments#3120 fix (or at least a workaround). Leaving this alone to make it easier for Azure / GitHub Actions team to see the problem.

Will open another PR without Ubuntu upgrade in a bit&hellip;

dougbu commented:
/fyi this may take a bit more time because

1. https://status.dev.azure.com/_event/235501110
2. Saw one test that likely won't be fixed by the environment variable workaround. Expect to need another bit from my #30729 internal branch.

dougbu committed:
Remove obsolete Groovy code

dougbu committed:
Pick up latest EF sources
- minor change shouldn't affect submodule build but best to keep up

dougbu commented on build/tools/docker/ubuntu.18.04/Dockerfile:
```diff
@@ -29,7 +30,8 @@ RUN apt-get update && \
         lldb-3.6 \
         wget && \
```
If this long list can be pared down because the Ubuntu 18.04 image contains them already, I'll do that in #30729

dougbu commented on src/Templating/test/Templates.Test/CdnScriptTagTests.cs:
```diff
@@ -96,6 +96,36 @@ public async Task FallbackSrcContent_Matches_CDNContent(ScriptTag scriptTag)
             Assert.Equal(RemoveLineEndings(cdnContent), RemoveLineEndings(fallbackSrcContent));
         }
 
+        class RetryHandler : DelegatingHandler
```
Surprisingly, it's the tests in src/Identity/test/Identity.Test/CdnScriptTaghelperTests.cs that have been flaky in this PR. The `RetryHelper` should be helpful in any case.

dougbu committed:
!fixup! More for 'Stop using Ubuntu 14.04'

dougbu commented:
/fyi I've been testing internal builds and addressing failures there. Test failures in PR validation builds appear to be flaky tests
- extra debug information for `IdentityUI_ScriptTags_SubresourceIntegrityCheck` tests didn't help much since SSL failure throws and `IdentityUI_ScriptTags_SubresourceIntegrityCheck.RetryHandler` doesn't catch the `Exception`
- Sockets.FunctionalTest/netcoreapp2.1 test group fails occasionally during teardown

dougbu commented:
Still blocked on Microsoft.AspNetCore.Identity.Test.CdnScriptTagTests.IdentityUI_ScriptTags_SubresourceIntegrityCheck

<details>

``` text

Error message
System.Net.Http.HttpRequestException : An error occurred while sending the request.
---- System.Net.WebException : The request was aborted: Could not create SSL/TLS secure channel.


Stack trace
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Identity.Test.CdnScriptTagTests.RetryHandler.<SendAsync>d__2.MoveNext() in /_/src/Identity/test/Identity.Test/CdnScriptTaghelperTests.cs:line 81
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.Http.HttpClient.<FinishSendAsyncUnbuffered>d__59.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.Http.HttpClient.<FinishGetStreamAsync>d__33.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Identity.Test.CdnScriptTagTests.<IdentityUI_ScriptTags_SubresourceIntegrityCheck>d__2.MoveNext() in /_/src/Identity/test/Identity.Test/CdnScriptTaghelperTests.cs:line 52
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
----- Inner Stack Trace -----
   at System.Net.HttpWebRequest.EndGetResponse(IAsyncResult asyncResult)
   at System.Net.Http.HttpClientHandler.GetResponseCallback(IAsyncResult ar)
```

</details>

I'm working to duplicate @javiercn's 3368ea6f9c4b / #13646 fix but can't get https://github.com/dotnet/aspnetcore/blob/release/3.1/src/Identity/test/Identity.Test/RetryHandler.cs#L57 to compile in this branch. Suspect updating https://github.com/dotnet/aspnetcore/blob/release/2.1/eng/targets/CSharp.Common.props#L4 to `8.0` or changing https://github.com/dotnet/aspnetcore/blob/release/2.1/src/Identity/test/Identity.Test/Microsoft.AspNetCore.Identity.Test.csproj to override that default would be sufficient. If increasing the C# language version is a breaking change, I'd do the second. Thoughts @dotnet/aspnet-build❔

dougbu commented:
/btw https://dev.azure.com/dnceng/internal/_build/results?buildId=1077719 confirmed our official builds (where we don't test 'natch) will succeed once this is in.

dougbu committed:
Back-port test reliability fix
- based on  @javiercn's 3368ea6 / #13646 fix
- avoid `TimeSpan` arithmetic and did not match refactoring done in 'release/3.1'
    - c# 8.0 is not supported in this branch

Pilchie commented:
@karelz @scalablecory - any ideas on the failure above in [Details](https://github.com/dotnet/aspnetcore/pull/31547#issuecomment-816182996).

It's coming from [here](https://github.com/dotnet/aspnetcore/blob/release/2.1/src/Identity/test/Identity.Test/CdnScriptTaghelperTests.cs#L51), and only seems to be failing in 2.1.

@dougbu I don't think you can update to compile with C# 8, because that requires a .NET SDK of 3.0 or later.  What is the error you are seeing?

dougbu commented:
> @dougbu I don't think you can update to compile with C# 8, because that requires a .NET SDK of 3.0 or later. What is the error you are seeing?

In local runs with C# 7.2, I saw

``` text
CdnScriptTaghelperTests.cs(105,51): error CS0019: Operator '*' cannot be applied to operands of type 'TimeSpan' and 'int' [...\src\Identity\test\Identity.Test\Microsoft.AspNetCore.Identity.Test.csproj]
```

To work around this, I ended up multiplying an `int` by `2` and creating a `TimeSpan` each time through the loop. See
https://github.com/dotnet/aspnetcore/pull/31547/files#diff-3b3c01f3096f00f0578c36dc3234ed37af00a870920fe3d79eafe8292f4cb47fR114
and
https://github.com/dotnet/aspnetcore/pull/31547/files#diff-3b3c01f3096f00f0578c36dc3234ed37af00a870920fe3d79eafe8292f4cb47fR139-R140

dougbu commented:
> only seems to be failing in 2.1

Successful test results don't show much information but could at least see success of `IdentityUI_ScriptTags_SubresourceIntegrityCheck` tests at https://dev.azure.com/dnceng/public/_build/results?buildId=1078781&view=ms.vss-test-web.build-test-results-tab. That's for 'main', where the test is a `[Theory]` and somewhat refactored but otherwise much like we have here.

Pilchie commented:
To be pendantic, that's not a C# compiler change, that's about the set of operators on `System.TimeSpan`.  A `static TimeSpan operator*(TimeSpan left, int right)` have been added after .NET Core 2.1.

dougbu commented:
Thanks for the additional context @Pilchie; I appreciate the pedantry.

----

Also asked whether this might be specific to the BuildPool.Server.Amd64.VS2017.Open queue in https://teams.microsoft.com/l/message/19:afba3d1545dd45d7b79f34c1821f6055@thread.skype/1617924854323?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=4d73664c-9f2f-450d-82a5-c2f02756606d&parentMessageId=1617924854323&teamName=.NET%20Core%20Eng%20Services%20Partners&channelName=First%20Responders&createdTime=1617924854323

dougbu commented:
@karelz @scalablecory @Pilchie 

Problem might be specific to .NET Framework runs and look relatively new. I saw it succeed on Monday in this branch (https://dev.azure.com/dnceng/public/_build/results?buildId=1073196) but it looked flaky in my internal builds for https://github.com/dotnet/aspnetcore/pull/30729 last week (https://dev.azure.com/dnceng/internal/_build?definitionId=21&branchFilter=150570%2C150570%2C150570%2C150570%2C150570%2C150570%2C150570%2C150570%2C150570%2C150570%2C150570%2C150570%2C150570%2C150570%2C150570%2C150570%2C150570%2C150570%2C150570).

I'm going to try skipping on .NET Framework. But, would very much like to know why this test went South.

dougbu committed:
Skip `CdnScriptTaghelperTests` on .NET Framework
- failures appear specific to that platform
- leave src/Templating/test/Templates.Test/CdnScriptTagTests.cs unchanged
    - test project does not target .NET Framework

dougbu commented on src/Identity/test/Identity.Test/CdnScriptTaghelperTests.cs:
```diff
@@ -23,7 +24,9 @@ public CdnScriptTagTests(ITestOutputHelper output)
-        [Fact]
+        // Because this test runs on .NET Core and seems reliable there, likely not worth running on .NET Framework.
+        [ConditionalFact]
+        [FrameworkSkipCondition(RuntimeFrameworks.CLR, SkipReason = "At least flaky on .NET Framework.")]
```
@Pilchie @dotnet/aspnet-build thoughts about filing a follow-up issue❔ I'm leaning toward no because the test is about external resources. Also realized the test runs fine on Linux and macOS.
![image](https://user-images.githubusercontent.com/6431421/114117089-cc01fd00-989a-11eb-8908-6fa359b21ab7.png)

wtgodbe replied:
I think we should at least file an issue about figuring out _why_ the test failed

wtgodbe replied:
If that's in dotnet/runtime, so be it

dougbu replied:
@wtgodbe I agree and wasn't really questioning that. My question was more whether we needed an issue about removing the `[FrameworkSkipCondition(...)]`. Hearing none 😃 

@karelz @scalablecory where should I file an issue about the .NET Framework `HttpClient` class and the underlying SSL libraries❔

karelz replied:
@dougbu I believe we do not run dotnet/runtime tests on .NET Framework anymore. Before we stopped to run them, we saw quite some flakiness on .NET Framework in networking. I wonder if this is similar ...
If you're able to isolate it to HttpClient/SslStream repro, we can take a look -- I would start email thread with our team as a starting point.

dougbu commented:
CI is green now 🎉 Just waiting for a response on adding an issue tracking restoring the skipped .NET Framework test. (Would link to the issue in the new `SkipReason`.)

----

Just checked and MusicStore.csproj targets only .NET Core in 'release/3.1' and 'release/5.0', which might explain why we haven't seen failures in those branches.
